helix-term/src/ui/picker.rs

```diff
@@ -56,22 +57,22 @@ pub const MAX_FILE_SIZE_FOR_PREVIEW: u64 = 10 * 1024 * 1024;
 #[derive(PartialEq, Eq, Hash)]
 pub enum PathOrId {
     Id(DocumentId),
-    Path(PathBuf),
+    Path(Arc<Path>),
```

Comment:
I am not sure that making this an Arc is quite the right thing to do.

A couple of points:

* most functions that create these have an `&PathBuf` that was cloned in the past. You then call `.into()` to turn it into an arc that is another allocation. It would be better to call `into()` directly
* However even that is inefficient since the result is cannoncalizaed and therefore turned back into a path buf which is then turned back into another arc (so another two allocations)

I think you could do this slightly differently: `PathOrId` and `FileLocation` are essentially just API types, they aren't stored anywhere just provided by callback functions. I think they really should be returning `&Path` which is more ergonomic in most places since the path is really just read from somewhere. 

Canonicalizing the path should also be done while inserting into the picker. Doing IO in the rendering loop really isn't ideal so that way those allocations also don't happen.

In the preview cache you already store an `Arc<Path>` as the key so you can use `cache.entry(path).key().clone()` to get an arc out (if the entry is occopuied if it isn't you can create a new arc from path an insert it so you would only need to create an arc one)